### PR TITLE
ICS3: fix inconsistencies in the spec and code of connection handshake

### PR DESCRIPTION
**Pull request by crodriguezvega**
This PR addresses the comments of #530 that were still relevant. I will copy those here:

- In connOpenInit() and connOpenTry(), there is no error handling in the [spec](https://github.com/cosmos/ics/tree/master/spec/ics-003-connection-semantics#opening-handshake) when calling addConnectionToClient(). In the [code](https://github.com/cosmos/cosmos-sdk/blob/fe8a891f11dcefe708e43d53549beec808f98687/x/ibc/core/03-connection/keeper/keeper.go#L158), the transaction is aborted if the client does not exist.

- In the code of connOpenTry(), but not in the spec: 
  - getting a consensus state from a given height [aborts on error](https://github.com/cosmos/cosmos-sdk/blob/fe8a891f11dcefe708e43d53549beec808f98687/x/ibc/core/03-connection/keeper/handshake.go#L93).

- In the code of connOpenAck(), but not in the spec: 
  - there is [error handling](https://github.com/cosmos/cosmos-sdk/blob/fe8a891f11dcefe708e43d53549beec808f98687/x/ibc/core/03-connection/keeper/handshake.go#L133) when getting the connection end by ID from the provable store.
  - getting a consensus state from a given height [aborts on error](https://github.com/cosmos/cosmos-sdk/blob/fe8a891f11dcefe708e43d53549beec808f98687/x/ibc/core/03-connection/keeper/handshake.go#L133).

- In the code of ConnOpenConfirm(), but not in the spec: there is [error handling](https://github.com/cosmos/cosmos-sdk/blob/fe8a891f11dcefe708e43d53549beec808f98687/x/ibc/core/03-connection/keeper/handshake.go#L133) when getting the connection end by ID from the provable store.

Besides that I also fixed some inconsistencies in the usage of the `versions` field in the connection end.

Closes: #530 
Closes: #894

### CODE REVIEW

File: spec/core/ics-003-connection-semantics/README.md [hunk 1]
```diff
@@ -125,6 +125,7 @@ function clientConnectionsPath(clientIdentifier: Identifier): Path {
 function addConnectionToClient(
   clientIdentifier: Identifier,
   connectionIdentifier: Identifier) {
+    abortTransactionUnless(queryClientState(clientIdentifier) !== null)
```
Comment (plafer) [inline spec/core/ics-003-connection-semantics/README.md:128]: IMO this is less clear than if it was a statement in the handlers' body directly ([as we do in ibc-rs](https://github.com/cosmos/ibc-rs/blob/1cbd1d93a87cd1967c943897d199f3274ee84259/crates/ibc/src/core/ics03_connection/handler/conn_open_init.rs#L24)). I also don't like that the check occurs after the `provableStore.set(connectionPath(identifier), connection)` line (e.g. in `ConnOpenTry`). Although I think we state somewhere that changes are reverted when a transaction is aborted, it's clearer to make all checks before state change statements.

Comment (crodriguezvega) [reply spec/core/ics-003-connection-semantics/README.md:128]: I see your point, thanks. I think I will defer to @AdityaSripal to decision to make such change. I would  also do it in a different PR that would close #894

**Review by AdityaSripal — APPROVED** [review]

File: spec/core/ics-003-connection-semantics/README.md (hunk 1 above)
Comment (AdityaSripal) [reply spec/core/ics-003-connection-semantics/README.md:128]: Don't understand the context here, please fill me in on slack @crodriguezvega

Comment (crodriguezvega) [reply spec/core/ics-003-connection-semantics/README.md]: Moved the statement to the body of `connOpenInit` and `connOpenTry`.

File: spec/core/ics-003-connection-semantics/README.md [hunk 2]
```diff
@@ -356,7 +360,7 @@ function connOpenInit(
     connection = ConnectionEnd{state, "", counterpartyPrefix,
       clientIdentifier, counterpartyClientIdentifier, versions, delayPeriodTime, delayPeriodBlocks}
     provableStore.set(connectionPath(identifier), connection)
-    addConnectionToClient(clientIdentifier, identifier)
+    addConnectionToClient(clientIdentifier, identifier) // aborts transaction if no client state is found
```
Comment (plafer) [inline spec/core/ics-003-connection-semantics/README.md:363]:
```suggestion
    addConnectionToClient(clientIdentifier, identifier)
```
removed comment as `addConnectionToClient` no longer aborts the transaction

File: spec/core/ics-003-connection-semantics/README.md [hunk 3]
```diff
@@ -382,24 +386,26 @@ function connOpenTry(
...
-                               clientIdentifier, counterpartyClientIdentifier, version, delayPeriodTime, delayPeriodBlocks}
+                               clientIdentifier, counterpartyClientIdentifier, [version], delayPeriodTime, delayPeriodBlocks}
     abortTransactionUnless(connection.verifyConnectionState(proofHeight, proofInit, counterpartyConnectionIdentifier, expectedConnectionEnd))
     abortTransactionUnless(connection.verifyClientState(proofHeight, proofClient, clientState))
     abortTransactionUnless(connection.verifyClientConsensusState(
       proofHeight, proofConsensus, counterpartyClientIdentifier, consensusHeight, expectedConsensusState))
 
     provableStore.set(connectionPath(identifier), connection)
-    addConnectionToClient(clientIdentifier, identifier)
+    addConnectionToClient(clientIdentifier, identifier) // aborts transaction if no client state is found
```
Comment (plafer) [inline spec/core/ics-003-connection-semantics/README.md:408]:
```suggestion
    addConnectionToClient(clientIdentifier, identifier)
```

**Review by plafer — APPROVED** [review]
Looks much better! Just need to remove the now incorrect comments

**Pull request merged**